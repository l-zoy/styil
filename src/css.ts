import { StyleSheetOptions } from './types'
import { createSelector } from './createSelector'
import { StyleSheet } from './sheet'
import { parseRules } from './parse'
import { CSSAttribute } from 'nativeCssTypes'

export function createStyil(options: Partial<StyleSheetOptions> = {}) {
  const { key = 'css', container, speedy, nonce } = options

  const globalCache = new Set<string>([])

  const sheet = new StyleSheet({
    key,
    speedy: speedy === undefined ? process.env.NODE_ENV === 'production' : speedy,
    container: globalThis.document ? container ?? globalThis.document.head : null,
    nonce
  })

  function createStyle(options: { styles: CSSAttribute; keyframes?: boolean; glob?: boolean }) {
    const { styles, keyframes, glob } = options
    const className = `${key}-${createSelector(styles)}`

    if (!globalCache.has(className)) {
      globalCache.add(className)

      sheet.insertStyle(
        parseRules(
          keyframes ? { [`@keyframes ${className}`]: styles } : styles,
          glob ? undefined : `.${className}`
        )
      )
    }

    return glob ? undefined : className
  }

  function css(styles: CSSAttribute) {
    return createStyle({ styles })
  }

  function glob(styles: CSSAttribute) {
    return createStyle({ styles, glob: true })
  }

  function keyframes(styles: CSSAttribute) {
    return createStyle({ styles, keyframes: true })
  }

  function getCssValue() {
    return sheet.ssrData
  }

  function flush() {
    sheet.flush()
    globalCache.clear()
  }

  return { css, glob, keyframes, getCssValue, flush }
}

export const { css, glob, keyframes, getCssValue, flush } = createStyil()
