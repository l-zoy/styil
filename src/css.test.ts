import { css, glob, keyframes, createStyils, flush } from './css'

describe('css', () => {
  afterEach(() => {
    flush()
  })

  it('base css', () => {
    const out = css({ foo: 1 })

    document.head.childNodes.forEach((item) => {
      expect(item.textContent).toEqual('.css-2165983570{foo:1px;}')
    })

    expect(out).toEqual('css-2165983570')
  })

  it('global css', () => {
    const out = glob({ body: { backgroundColor: 'red' } })
    document.head.childNodes.forEach((item) => {
      expect(item.textContent).toEqual('body{background-color:red;}')
    })

    expect(out).toEqual(undefined)
  })

  it('keyframes css', () => {
    const out = keyframes({
      from: {
        transform: 'rotate(0deg)'
      },
      to: {
        transform: 'rotate(360deg)'
      }
    })

    document.head.childNodes.forEach((item) => {
      expect(item.textContent).toEqual(
        '@keyframes css-524936959{from{transform:rotate(0deg);}to{transform:rotate(360deg);}}'
      )
    })

    expect(out).toEqual('css-524936959')
  })

  it('createStyils', () => {
    const { css, flush } = createStyils({
      key: 'abc',
      container: document.body,
      nonce: '123456'
    })
    const out = css({
      width: 'auto'
    })

    expect(document.documentElement).toMatchSnapshot()

    expect(out).toEqual('abc-3169980522')
    flush()
  })

  it('createStyils speedy', () => {
    process.env.NODE_ENV = 'production'
    const { css } = createStyils()
    const out = css({
      width: 'auto',
      '.foo': {
        height: 'auto'
      }
    })

    expect(document.documentElement).toMatchSnapshot()

    expect(out).toEqual('css-3676222078')
  })
})
