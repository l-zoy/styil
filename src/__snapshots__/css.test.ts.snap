// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css createStyil 1`] = `
<html>
  <head />
  <body>
    <style
      data-styils="abc"
      nonce="123456"
    >
      
      .abc-2883881200{width:auto;}
    </style>
  </body>
</html>
`;

exports[`css createStyil speedy 1`] = `
<html>
  <head>
    <style
      data-styils="css"
    >
      
    </style>
  </head>
  <body />
</html>
`;
